plugin/src/main/kotlin/tanvd/grazi/language/LangDownloader.kt

```diff
@@ -0,0 +1,174 @@
+                    if (exception != null) {
+                        throw RuntimeException(exception)
+                    }
+                }
+            })
+
+            getService(RepositorySystem::class.java)
+        }
+    }
+
+    private val session: RepositorySystemSession by lazy {
+        with(MavenRepositorySystemUtils.newSession()) {
+            localRepositoryManager = repository.newLocalRepositoryManager(this, LocalRepository(File(GraziPlugin.installationFolder, "poms")))
+            setProxySelector(JreProxySelector())
+        }
+    }
+
+    private val proxy = JreProxySelector()
+
+    private val MAVEN_CENTRAL_REPOSITORY =
+            RemoteRepository.Builder("central", "default", msg("grazi.maven.repo.url"))
+                    .setProxy(proxy.getProxy(msg("grazi.maven.repo.url"))).build()
+
+    private fun Artifact.createDependency() = Dependency(this, JavaScopes.COMPILE, false, listOf(
+            Exclusion("org.languagetool", "languagetool-core", "", "jar"),
+            Exclusion("org.carrot2", "morfologik-fsa", "", "jar"),
+            Exclusion("org.carrot2", "morfologik-stemming", "", "jar"),
+            Exclusion("com.google.guava", "guava", "", "jar")
+    ))
+
+    private val Artifact.name
+        get() = "$artifactId-$version.jar"
+
+    private val Artifact.url
+        get() = "${MAVEN_CENTRAL_REPOSITORY.url}${groupId.replace(".", "/")}/$artifactId/$version/$name"
+
+    private val downloader by lazy { DownloadableFileService.getInstance() }
+
+    private fun DependencyNode.traverse(action: (DependencyNode) -> Unit) {
+        action(this)
+        this.children.forEach(action)
+    }
+
+    fun downloadMissingLanguages(project: Project?) {
+        val state = GraziConfig.get()
+
+        if (state.hasMissedLanguages()) {
+            state.enabledLanguages.filter { it.jLanguage == null }.forEach {
+                with(LangDownloader) { it.downloadLanguage(project) }
+            }
+
+            if (state.nativeLanguage.jLanguage == null) {
+                with(LangDownloader) { state.nativeLanguage.downloadLanguage(project) }
+            }
+
+            ProjectManager.getInstance().openProjects.forEach {
+                DaemonCodeAnalyzer.getInstance(it).restart()
+            }
+        }
+    }
+
+    private fun Lang.registerInLanguageTool() {
+        with(Languages::class.java.getDeclaredField("dynLanguages")) {
+            isAccessible = true
+            @Suppress("UNCHECKED_CAST")
+            val langs = get(null) as MutableList<Language>
+
+            langs.add(jLanguage!!)
+            when (this@registerInLanguageTool) {
+                Lang.BRAZILIAN_PORTUGUESE -> langs.add(Lang.PORTUGAL_PORTUGUESE.jLanguage!!)
+                Lang.PORTUGAL_PORTUGUESE -> langs.add(Lang.BRAZILIAN_PORTUGUESE.jLanguage!!)
+                Lang.GERMANY_GERMAN -> langs.add(Lang.AUSTRIAN_GERMAN.jLanguage!!)
```

Comment:
Hm, it is better to define dialects that should be registered at once. Some `LangToolLibDescriptor` with name and dialects that it provide?

Reply:
Done